Reqnroll.Tools.MsBuild.Generation/CodeBehindWriter.cs

```diff
@@ -42,10 +42,36 @@ public CodeBehindWriter(TaskLoggingHelper log)
                     Directory.CreateDirectory(directoryPath);
                 }
 
-                File.WriteAllText(outputPath, testFileGeneratorResult.GeneratedTestCode, Encoding.UTF8);
+                WriteAllTextWithRetry(outputPath, testFileGeneratorResult.GeneratedTestCode, Encoding.UTF8);
             }
 
             return outputPath;
         }
+
+        /// <summary>
+        /// When building a multi-targeted project, the build system may try to write the same file multiple times,
+        /// and this can cause an IOException ("The process cannot access the file because it is being used by another process.").
+        /// See https://github.com/reqnroll/Reqnroll/issues/197
+        /// Once we move to Roslyn-based generation, this problem will go away, but for now, we use a workaround of
+        /// retrying the write operation a few times (the content is anyway the same).
+        /// </summary>
+        private void WriteAllTextWithRetry(string path, string contents, Encoding encoding)
+        {
+            const int maxAttempts = 5;
+            for (int i = 1; i <= maxAttempts; i++)
+            {
+                try
+                {
+                    File.WriteAllText(path, contents, encoding);
+                    return;
+                }
+                catch (IOException)
+                {
+                    if (i == maxAttempts)
+                        throw;
+                    System.Threading.Thread.Sleep(i * 50);
+                }
+            }
+        }
```

Comment:
Just wanted to add: genuinely surprised there isn't a better solution for this in 2024 than a try-catch-retry loop, but I couldn't find anything that would obviously work better and must conclude it is The Way. 😅

Reply:
Yeah, I agree. Sad.